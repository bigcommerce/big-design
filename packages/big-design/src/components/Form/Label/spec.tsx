import { render, screen } from '@testing-library/react';
import React from 'react';

import 'jest-styled-components';

import { FormControlLabel } from './index';

test('renders a label', () => {
  render(<FormControlLabel htmlFor="test-id">This is a label</FormControlLabel>);

  expect(screen.getByText('This is a label')).toBeInTheDocument();
});

test('does not forward styles', () => {
  render(
    <FormControlLabel className="test" htmlFor="test-id" style={{ background: 'red' }}>
      Label
    </FormControlLabel>,
  );

  expect(screen.queryByText('test')).not.toBeInTheDocument();
  expect(screen.getByText('Label')).not.toHaveStyle('background: red');
});

test('appends * text when required', () => {
  render(
    <FormControlLabel htmlFor="test-id" required>
      This is a label
    </FormControlLabel>,
  );

  expect(screen.getByText('This is a label')).toBeInTheDocument();
  expect(screen.getByText('*')).toBeInTheDocument();
});
