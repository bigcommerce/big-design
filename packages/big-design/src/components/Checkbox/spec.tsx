import { render, screen } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import React, { createRef } from 'react';
import 'jest-styled-components';

import { warning } from '../../utils';

import { CheckboxLabel } from './Label';

import { Checkbox } from './index';

describe('render Checkbox', () => {
  test('checked', () => {
    render(<Checkbox checked={true} label="Checked" onChange={() => null} />);

    expect(screen.getByRole('checkbox')).toMatchSnapshot();
  });

  test('unchecked', () => {
    render(<Checkbox checked={false} label="Unchecked" onChange={() => null} />);

    expect(screen.getByRole('checkbox')).toMatchSnapshot();
  });

  test('with description object', () => {
    render(
      <Checkbox
        checked={false}
        description={{
          text: 'description',
          link: {
            text: 'learn more',
            target: 'foo',
            href: 'bar',
          },
        }}
        label="Unchecked"
        name="test-group"
        onChange={() => null}
      />,
    );

    expect(screen.getByRole('checkbox')).toMatchSnapshot();
  });

  test('with badge props', () => {
    render(
      <Checkbox
        badge={{
          label: 'Badge',
          variant: 'secondary',
        }}
        description="description text"
        label="Unchecked"
        name="test-group"
        onChange={() => null}
      />,
    );

    expect(screen.getByRole('checkbox')).toBeInTheDocument();
  });

  test('with description string', () => {
    render(
      <Checkbox
        checked={false}
        description="description text"
        label="Unchecked"
        name="test-group"
        onChange={() => null}
      />,
    );

    expect(screen.getByRole('checkbox')).toMatchSnapshot();
  });

  test('indeterminate', () => {
    render(<Checkbox isIndeterminate={true} label="Unchecked" onChange={() => null} />);

    expect(screen.getByRole('checkbox')).toMatchSnapshot();
  });

  test('disabled checked', () => {
    render(<Checkbox checked={true} disabled label="Checked" onChange={() => null} />);

    expect(screen.getByRole('checkbox')).toMatchSnapshot();
  });

  test('disabled unchecked', () => {
    render(<Checkbox checked={false} disabled label="Checked" onChange={() => null} />);

    expect(screen.getByRole('checkbox')).toMatchSnapshot();
  });

  test('disabled indeterminate', () => {
    render(<Checkbox disabled isIndeterminate={true} label="Unchecked" onChange={() => null} />);

    expect(screen.getByRole('checkbox')).toMatchSnapshot();
  });
});

test('has correct value for checked', async () => {
  render(<Checkbox checked={true} data-testid="checkbox" label="Checked" onChange={() => null} />);

  const input = await screen.findByTestId<HTMLInputElement>('checkbox');

  expect(input).toBeChecked();
});

test('has correct value for unchecked', async () => {
  render(<Checkbox checked={false} data-testid="checkbox" label="Checked" onChange={() => null} />);

  const input = await screen.findByTestId<HTMLInputElement>('checkbox');

  expect(input).not.toBeChecked();
});

test('has correct value for indeterminate', async () => {
  render(
    <Checkbox
      checked={false}
      data-testid="checkbox"
      isIndeterminate
      label="Checked"
      onChange={() => null}
    />,
  );

  const input = await screen.findByTestId<HTMLInputElement>('checkbox');

  expect(input).not.toBeChecked();
});

test('triggers onChange when clicking the checkbox', async () => {
  const onChange = jest.fn();

  render(<Checkbox checked={true} data-testid="checkbox" label="Checked" onChange={onChange} />);

  const checkbox = await screen.findByTestId<HTMLInputElement>('checkbox');

  await userEvent.click(checkbox);

  expect(onChange).toHaveBeenCalled();
});

test('triggers onChange when clicking styled and text label', async () => {
  const onChange = jest.fn();

  render(<Checkbox checked={true} data-testid="checkbox" label="Checked" onChange={onChange} />);

  const labels = screen.getAllByText('Checked');

  await userEvent.click(labels[0]);

  expect(onChange).toHaveBeenCalled();
});

test('accepts valid CheckboxLabel component', () => {
  const testId = 'test';
  const label = <CheckboxLabel data-testid={testId}>Label</CheckboxLabel>;

  render(<Checkbox label={label} />);

  expect(screen.getByTestId(testId)).toBeInTheDocument();
});

test('does not accept invalid label component', () => {
  const testId = 'test';
  const label = <div data-testid={testId}>Label</div>;

  render(<Checkbox label={label} />);

  expect(warning).toHaveBeenCalledTimes(1);
  expect(screen.queryByTestId(testId)).not.toBeInTheDocument();
});

test('forwards ref', () => {
  const ref = createRef<HTMLInputElement>();

  render(<Checkbox data-testid="checkbox" label="Checked" ref={ref} />);

  const input = screen.getByTestId('checkbox');

  expect(input).toBe(ref.current);
});

test('does not forward styles', () => {
  render(<Checkbox className="test" label="Checked" style={{ background: 'red' }} />);

  expect(screen.queryByRole('checkbox', { name: /test/i })).not.toBeInTheDocument();
  expect(screen.getByRole('checkbox')).not.toHaveStyle('background: red');
});
