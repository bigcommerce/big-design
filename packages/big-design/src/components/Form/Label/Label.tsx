import React, { ComponentPropsWithoutRef } from 'react';

import { Text } from '../../Typography/';

import { StyledLabel } from './styled';

export interface FormControlLabelProps extends ComponentPropsWithoutRef<'label'> {
  required?: boolean;
}

export const FormControlLabel: React.FC<FormControlLabelProps> = ({
  className,
  style,
  required,
  ...props
}) => (
  <StyledLabel {...props}>
    {props.children}
    {required && (
      <Text aria-hidden="true" as="span" color="danger">
        {' '}
        *
      </Text>
    )}
  </StyledLabel>
);
