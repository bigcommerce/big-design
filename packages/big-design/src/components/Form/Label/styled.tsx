import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import styled, { DefaultTheme, StyledComponent } from 'styled-components';

import { StyleableH4 } from '../../Typography/private';
import { HeadingProps } from '../../Typography/types';

import { type FormControlLabelProps } from './Label';

interface StyledLabelArgument extends Omit<FormControlLabelProps, 'localization'> {
  theme: DefaultTheme;
}

export const StyledLabel = styled<
  StyledComponent<'label' | 'h4', DefaultTheme, Partial<HeadingProps>> & FormControlLabelProps
>(StyleableH4).attrs({
  as: 'label',
})<FormControlLabelProps>`
  cursor: pointer;
  display: inline-block;
  margin-bottom: ${({ theme }: StyledLabelArgument) => theme.spacing.xxSmall};
`;

StyledLabel.defaultProps = { theme: defaultTheme };
