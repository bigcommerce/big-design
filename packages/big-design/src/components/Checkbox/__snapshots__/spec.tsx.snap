// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`render Checkbox checked 1`] = `
.c0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<input
  aria-checked="true"
  aria-labelledby=":r1:"
  checked=""
  class="c0"
  id=":r0:"
  type="checkbox"
/>
`;

exports[`render Checkbox disabled checked 1`] = `
.c0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<input
  aria-checked="true"
  aria-labelledby=":rj:"
  checked=""
  class="c0"
  disabled=""
  id=":ri:"
  type="checkbox"
/>
`;

exports[`render Checkbox disabled indeterminate 1`] = `
.c0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<input
  aria-labelledby=":rp:"
  class="c0"
  disabled=""
  id=":ro:"
  type="checkbox"
/>
`;

exports[`render Checkbox disabled unchecked 1`] = `
.c0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<input
  aria-checked="false"
  aria-labelledby=":rm:"
  class="c0"
  disabled=""
  id=":rl:"
  type="checkbox"
/>
`;

exports[`render Checkbox indeterminate 1`] = `
.c0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<input
  aria-labelledby=":rg:"
  class="c0"
  id=":rf:"
  type="checkbox"
/>
`;

exports[`render Checkbox unchecked 1`] = `
.c0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<input
  aria-checked="false"
  aria-labelledby=":r4:"
  class="c0"
  id=":r3:"
  type="checkbox"
/>
`;

exports[`render Checkbox with description object 1`] = `
.c0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<input
  aria-checked="false"
  aria-labelledby=":r7:"
  class="c0"
  id=":r6:"
  name="test-group"
  type="checkbox"
/>
`;

exports[`render Checkbox with description string 1`] = `
.c0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<input
  aria-checked="false"
  aria-labelledby=":rd:"
  class="c0"
  id=":rc:"
  name="test-group"
  type="checkbox"
/>
`;
